configurable keyboard shortcuts

umlaeute opened the pull request:
this PR makes keyboard shortcuts configurable by the user.
- Closes: https://github.com/pure-data/pure-data/issues/2353
- Closes: https://github.com/pure-data/pure-data/issues/1729
- Closes: https://github.com/pure-data/pure-data/issues/2349
- Closes: https://github.com/pure-data/pure-data/issues/1914

So if you prefer <kbd>Control</kbd>+<kbd>space</kbd> to save your documents or quit Pd, you can have that.
Or if you think that you would rather have a shortcut for the `Audio settings` than for `Find again`, be my friend.
While testing, I found <kbd>Control</kbd>+<kbd>,</kbd> to open up the preferences very useful :-)

### under the hood
in order to synchronize the keyboard accelerators keys displayed in the menu bar (which doesn't do anything) with the real shortcuts, this PR creates a new indirection:
- each menu action is now triggered by a virtual event. e.g. for saving a file, you can send the `<<File|Save>>` event to a patch window
- each virtual event can be associated with a keyboard shortcut (or multiple keyboard shortcuts)
- each menu entry only triggers the corresponding virtual event
- the accelerator key (shown in the menu) is calculated from the sequence associated with the virtual event

e.g. the `<<File|New>>` event triggers the execution of the `menu_new` command to create a new patch.
if you use a specific shortcut (e.g. <kbd>Control</kbd>+<kbd>n</kbd>) the associated virtual event (here `<<File|New>>`) is sent to the current window, this creating a new patch.
if otoh you navigate the menu to `File` -> `New`, you will trigger the same virtual event (`<<File|New>>`), with the same effect.
when the menu associated with the `<<File|New>>` event is created (or updated), it looks up the shortcut sequence for this event (which in tcl is `Control-Key-n`) and translates it to a visually pleasing acceleration string (e.g. `Ctrl-N`)


### editing sequences
the preference window gained a new tab "keyboard shortcuts", where you can interactively record a new shortcut for each event.
**TODO** as of now, the mapping goes from "virtual event" (as opposed to "menu entry") to the shortcut, which gives slightly raw strings (e.g. `Find|FindLastError|` is displayed as `Find`->`FindLastError`) which shouldn't really be translated. however, i think we can come up with something that uses the menu labels)

### issues
- [x] no keybindings for <kbd>Command</kbd> and <kbd>Option</kbd> (required for macOS)
- [x] (re)storing preferences on macOS does not work (see also #2381)
- [x] check for duplicates does not work reliably (e.g. <kbd>Command</kbd>+<kbd>Shift</kbd>+<kbd>S</kbd> is not detected to be the same as <kbd>Shift</kbd>+<kbd>Command</kbd>+<kbd>S</kbd>)
- [x] printing a patch always suggest to print as `ERROR.ps`
- [x] on macOS, you have to release the terminal key first when recording a new shortcut (e.g. recording <kbd>Command</kbd>+<kbd>Shift</kbd>+<kbd>R</kbd>, you start with pressing <kbd>Command</kbd>+<kbd>Shift</kbd>, holding these and then press <kbd>R</kbd>; if you then release the <kbd>Shift</kbd>-key first,, followed by <kbd>R</kbd> the recording stays in limbo; if you instead release the <kbd>R</kbd>-key first, everything works as expected)
- [ ] virtual events should also be used in in dialogs
- [ ] shortcut editor could be more intuitive
- [ ] shortcut editor sometimes does not show the currently pressed keys

### test binaries
- [macOS universal](https://git.iem.at/zmoelnig/pure-data/-/jobs/artifacts/feature/keyboard-shortcuts/download?job=macOS)
- [Windows 64](https://git.iem.at/zmoelnig/pure-data/-/jobs/artifacts/feature/keyboard-shortcuts/download?job=w64)
- [Windows 32](https://git.iem.at/zmoelnig/pure-data/-/jobs/artifacts/feature/keyboard-shortcuts/download?job=w32)


if the links above do not work, you should be able to find artifacts at <https://git.iem.at/zmoelnig/pure-data/-/pipelines?ref=feature%2Fkeyboard-shortcuts> (somewhere at the right-hand side  there's a download pulldown)

umlaeute committed:
pd_bindings: define "alt" and "control" variables for the bindings

umlaeute committed:
pd_bindings: use virtual events for menus

- pass all events through ::pd_menucommands::scheduleAction
- new virtual event categories "Preferences" and "Pd"
  - "Preferences" is obvious (and contain audio/midi settings)
  - "Pd" is a bit of a catchall: e.g. sending messages to Pd has nothing
    to do with "File" operations

- explicitely initialize pd_bindings via ::pd_bindings::setup

umlaeute committed:
switch menu-commands to virtual events

use helper proc for creating menus

umlaeute committed:
pd_bindings: try to read key bindings from GUI-prefs

the foundation for https://github.com/pure-data/pure-data/issues/2353

we still need an easy way to change the keybindings

umlaeute committed:
shortcut editor

umlaeute committed:
pd_bindings: split out the actual event <- shortcut binding from pd_binding::setup

umlaeute committed:
proc to update all accelerators in menus

umlaeute committed:
menus: proc to get virtual events triggered directly from menu

porres commented:
can't download the test binary for macOS

<img width="843" alt="Screenshot 2024-07-11 at 16 23 26" src="https://github.com/user-attachments/assets/c769e5ba-9861-489c-8734-25913f47e80a">

umlaeute commented:
it's now working again.
i'm not sure why these URLs are unstable.

in any case, if the fail again in the future, you should be able to find artifacts at <https://git.iem.at/pd/pure-data/-/pipelines?ref=feature%2Fkeyboard-shortcuts>

porres commented:
cool, not sure yet how to edit these shortcuts on the Preference's tab

umlaeute commented:
there's an obvious bug on macOS, where the stored KeyBindings cannot be properly read again. 

if you get into errors when opening the Preferences (again), you can manually delete the offending setting on the cmdline:

```sh
defaults delete org.puredata.pd.pd-gui KeyBindings
```


here, storing the default bindings writes somethings like the following to the plist:
```
File|New \\{\\{Command N\\}\\} File|Open \\{\\{Command O\\}\\} File|Save \\{\\{Command S\\}\\} File|SaveAs \\{\\{Command Shift S\\}\\} File|Print \\{\\{Command P\\}\\} File|ClearRecentFiles \\{\\} File|Close \\{\\{Command W\\}\\} File|CloseNow \\{\\{Command Shift W\\}\\} File|QuitNow \\{\\{Command Shift Q\\}\\} Edit|Undo \\{\\{Command Z\\}\\} Edit|Redo \\{\\{Command Shift Z\\}\\} Edit|Cut \\{\\{Command X\\}\\} Edit|Copy \\{\\{Command C\\}\\} Edit|Paste \\{\\{Command V\\}\\} Edit|PasteReplace \\{\\} Edit|Duplicate \\{\\{Command D\\}\\} Edit|SelectAll \\{\\{Command A\\}\\} Edit|SelectNone Escape Edit|Font \\{\\} Edit|ZoomIn \\{\\{Command plus\\} \\{Command KP_Add\\} \\{Command equal\\}\\} Edit|ZoomOut \\{\\{Command minus\\} \\{Command KP_Subtract\\}\\} Edit|TidyUp \\{\\{Command Shift R\\}\\} Edit|ConnectSelection \\{\\{Command K\\}\\} Edit|Triggerize \\{\\{Command T\\}\\} Edit|EditMode \\{\\{Command E\\}\\} Put|Object \\{\\{Command 1\\}\\} Put|Message \\{\\{Command 2\\}\\} Put|Number \\{\\{Command 3\\}\\} Put|List \\{\\{Command 4\\}\\} Put|Symbol \\{\\} Put|Comment \\{\\{Command 5\\}\\} Put|Bang \\{\\{Command Shift B\\}\\} Put|Toggle \\{\\{Command Shift T\\}\\} Put|Number2 \\{\\{Command Shift N\\}\\} Put|VerticalSlider \\{\\{Command Shift V\\}\\} Put|HorizontalSlider \\{\\{Command Shift J\\}\\} Put|VerticalRadio \\{\\{Command Shift D\\}\\} Put|HorizontalRadio \\{\\{Command Shift I\\}\\} Put|VUMeter \\{\\{Command Shift U\\}\\} Put|Canvas \\{\\{Command Shift C\\}\\} Put|Graph \\{\\{Command Shift G\\}\\} Put|Array \\{\\{Command Shift A\\}\\} Find|Find \\{\\{Command F\\}\\} Find|FindAgain \\{\\{Command G\\}\\} Find|FindLastError \\{\\} Media|DSPOn \\{\\{Command slash\\}\\} Media|DSPOff \\{\\{Command period\\}\\} Media|TestAudioMIDI \\{\\} Media|LoadMeter \\{\\} Window|Maximize \\{\\} Window|AllToFront \\{\\} Window|PdWindow \\{\\{Command R\\}\\} Window|Parent \\{\\} Help|About \\{\\} Help|Manual \\{\\} Help|Browser \\{\\{Command B\\}\\} Help|ListObjects \\{\\} Help|puredata.info \\{\\} Help|CheckUpdates \\{\\} Help|ReportBug \\{\\} Pd|Message \\{\\{Command Shift M\\}\\} Pd|ClearConsole \\{\\{Command Shift L\\}\\} Preferences|Audio \\{\\} Preferences|MIDI \\{\\} Preferences|Edit \\{\\} Preferences|Save \\{\\} Preferences|SaveTo \\{\\} Preferences|Load \\{\\} Preferences|Forget \\{\\}
```

which is then read back as 
```
File|New {\{\{Command} {N\}\}} File|Open {\{\{Command} {O\}\}} File|Save {\{\{Command} {S\}\}} File|SaveAs {\{\{Command} Shift {S\}\}} File|Print {\{\{Command} {P\}\}} File|ClearRecentFiles {\{\}} File|Close {\{\{Command} {W\}\}} File|CloseNow {\{\{Command} Shift {W\}\}} File|QuitNow {\{\{Command} Shift {Q\}\}} Edit|Undo {\{\{Command} {Z\}\}} Edit|Redo {\{\{Command} Shift {Z\}\}} Edit|Cut {\{\{Command} {X\}\}} Edit|Copy {\{\{Command} {C\}\}} Edit|Paste {\{\{Command} {V\}\}} Edit|PasteReplace {\{\}} Edit|Duplicate {\{\{Command} {D\}\}} Edit|SelectAll {\{\{Command} {A\}\}} Edit|SelectNone Escape Edit|Font {\{\}} Edit|ZoomIn {\{\{Command} {plus\}} {\{Command} {KP_Add\}} {\{Command} {equal\}\}} Edit|ZoomOut {\{\{Command} {minus\}} {\{Command} {KP_Subtract\}\}} Edit|TidyUp {\{\{Command} Shift {R\}\}} Edit|ConnectSelection {\{\{Command} {K\}\}} Edit|Triggerize {\{\{Command} {T\}\}} Edit|EditMode {\{\{Command} {E\}\}} Put|Object {\{\{Command} {1\}\}} Put|Message {\{\{Command} {2\}\}} Put|Number {\{\{Command} {3\}\}} Put|List {\{\{Command} {4\}\}} Put|Symbol {\{\}} Put|Comment {\{\{Command} {5\}\}} Put|Bang {\{\{Command} Shift {B\}\}} Put|Toggle {\{\{Command} Shift {T\}\}} Put|Number2 {\{\{Command} Shift {N\}\}} Put|VerticalSlider {\{\{Command} Shift {V\}\}} Put|HorizontalSlider {\{\{Command} Shift {J\}\}} Put|VerticalRadio {\{\{Command} Shift {D\}\}} Put|HorizontalRadio {\{\{Command} Shift {I\}\}} Put|VUMeter {\{\{Command} Shift {U\}\}} Put|Canvas {\{\{Command} Shift {C\}\}} Put|Graph {\{\{Command} Shift {G\}\}} Put|Array {\{\{Command} Shift {A\}\}} Find|Find {\{\{Command} {F\}\}} Find|FindAgain {\{\{Command} {G\}\}} Find|FindLastError {\{\}} Media|DSPOn {\{\{Command} {slash\}\}} Media|DSPOff {\{\{Command} {period\}\}} Media|TestAudioMIDI {\{\}} Media|LoadMeter {\{\}} Window|Maximize {\{\}} Window|AllToFront {\{\}} Window|PdWindow {\{\{Command} {R\}\}} Window|Parent {\{\}} Help|About {\{\}} Help|Manual {\{\}} Help|Browser {\{\{Command} {B\}\}} Help|ListObjects {\{\}} Help|puredata.info {\{\}} Help|CheckUpdates {\{\}} Help|ReportBug {\{\}} Pd|Message {\{\{Command} Shift {M\}\}} Pd|ClearConsole {\{\{Command} Shift {L\}\}} Preferences|Audio {\{\}} Preferences|MIDI {\{\}} Preferences|Edit {\{\}} Preferences|Save {\{\}} Preferences|SaveTo {\{\}} Preferences|Load {\{\}} Preferences|Forget {\{\}} Preferences|Edit {} Preferences|Save {} Preferences|SaveTo {} Preferences|Load {} Preferences|Forget {} File|Open {} File|SaveAs {} Pd|Message {} File|Print {} Edit|Undo {} Edit|Copy {} Edit|Duplicate {} Edit|PasteReplace {} Edit|Font {} Edit|ZoomIn {} Edit|TidyUp {} Edit|ConnectSelection {} Pd|ClearConsole {} Put|Message {} Put|List {} Put|Bang {} Put|Toggle {} Put|Number2 {} Put|VerticalSlider {} Put|HorizontalSlider {} Put|VerticalRadio {} Put|HorizontalRadio {} Put|VUMeter {} Put|Canvas {} Put|Graph {} Put|Array {} Find|Find {} Find|FindLastError {} Media|DSPOn {} Media|TestAudioMIDI {} Media|LoadMeter {} Preferences|Audio {} Preferences|MIDI {} Window|PdWindow {} Help|ListObjects {} Help|puredata.info {} Help|CheckUpdates {} Help|ReportBug {}
```

yikes :-(

the other problem seems to be, that (at least with my VM), the <kbd>⌘</kbd> key is not recognized

umlaeute commented:
the other problem seems to be, that (at least with my VM), the <kbd>⌘</kbd> key is not recognized

umlaeute commented:
afaict, i've fixed the two problems mentioned above for macOS (as of 33230bfa312fd4d1a4e05d9146cd2fb2cd26cba1)

so please test (again).

@porres wrote:
> cool, not sure yet how to edit these shortcuts on the Preference's tab

ideally you should just follow the on-screen instructions:
```
To edit a keyboard shortcut, doubleclick on the corresponding row and type a new accelerator, or press BackSpace to clear.
```

let me know, if something is unclear.

(sidenote: afaict, there's no way to override apple's built-in shortcuts, e.g. <kbd>Command</kbd>+<kbd>,</kbd> will *always* give you the preferences (even within the shortcut editor). the same goes for <kbd>Command</kbd>+<kbd>q</kbd> for quitting)

porres commented:
> EDIT this should be fixed by now.
> (i forgot to bind ⌘ and Option to the shortcut editor)

Are the binaries updated for testing?

porres commented:
hey, ok, testing it right now, seems much better and intuitive now, cool

porres commented:
now, hey, why are there like 3 options for shortcuts?

umlaeute commented:
> now, hey, why are there like 3 options for shortcuts?

so you can add multiple shortcuts to a single action.
you might have noticed, that e.g. `Zoom In` is already bound to <kbd>Control</kbd>+<kbd>=</kbd>,  <kbd>Control</kbd>+<kbd>+</kbd> and <kbd>Control</kbd>+<kbd>+</kbd> (NumPad).

umlaeute commented:
> hey, ok, testing it right now, seems much better and intuitive now, cool

sidenote: it was like this in the previous iteration as well (just that the `Command` and `Option` keys didn't do anything, which made the editor a bit useless on macOS)

porres commented:
The instructions need to say "To edit a shortcut, doubleclick on the corresponding ROW AND COLUMN", or "cell", or something. Just "row" is wrong and misleading.

porres commented:
Also, I think it's better is the instructions are on top instead of the bottom area

porres commented:
A DREAM COME TRUE!!! 

<img width="334" alt="Screenshot 2024-07-14 at 15 02 19" src="https://github.com/user-attachments/assets/270a5cda-063c-491d-bc0a-6fbe91394f88">

umlaeute commented:
> The instructions need to say "To edit a shortcut, doubleclick on the corresponding ROW AND COLUMN", or "cell", or something. Just "row" is wrong and misleading.

it's not wrong.

clicking anywhere on the row, will allow you add/update a shortcut ("update" if it is a cell, "add" otherwise).
i think that speaking of "cells" or "rows and columns" is rather confusing (esp. since the columns are not visibly distinguished).

i had the help-text on top first, but moved it to the bottom after checking with TheGIMP. i don't have much opinion about it's position though.

porres commented:
hmmm, I see that double clicking on "event" also works.

Anyway, I think that a full and throughly description on how this works should be in the manual anyway

I see that double clicking on an event and typing something creates a new #1 shortcut and moves the previous one to #2... 

To edit it properly one needs to double click the cell.

These differences should be explained on the manual

Anyway, I still feel things are either unclear, not fully explained or wrong in the current brief instruction. Like, if I click on the row and use basckspace, the shortcut isn't deleted as the instructions would have me believe... I need to select the exact 'cell' for that.

umlaeute commented:
> Anyway, I still feel things are either unclear, not fully explained or wrong in the current brief instruction. Like, if I click on the row and use basckspace, the shortcut isn't deleted as the instructions would have me believe... I need to select the exact 'cell' for that.

here (tested on Linux and macOS/sierra), clicking on either a row or a cell doesn't delete a shortcut.
however, double-clicking (as mentioned in the brief instruction) on a *cell* and using <kbd>BackSpace</kbd> deletes the specific shortcut, and double-clicking on a *row* (that is: the event) and using <kbd>BackSpace</kbd> deletes *all* shortcuts for the given event - as I would have expected it.

personally, i would prefer if people can use the shortcut editor without having to resort to the manual..
so it should be intuitive.
i'm obviously a bit biased what i think is intuitive.

rather than having more personal suggestions on what might be intuitive, i would like to model the behavior after applications that already use a shortcut editor successfully (and where i do not have to read the manual to use the editor).
what works for me:
- The GIMP
- Inkscape

what does not work for me, is
- ~~LibreOffice~~

happy to hear of other applications (if they are non-free, do not hesitate to add screenshots)

ben-wes commented:
thanks a lot for these additions - cool!

personally, i admit that i barely ever change shortcuts for any application since i tend to hope for good defaults (believing that there's some reasoning behind them that makes it worth learning these).

some feedback on this (macbook air m2, "German - Standard" keyboard layout and input source):
* no scrollbar visible for me
* touchpad movements stop the "recording" when the cursor is moved out off the recorded cell
* the shortcut text is clipped a bit much in the cell while recording with fontsize 10
* modifiers appear as 2 keys while recording, like "Command+Meta_L" (i guess, this is a proper representation - but it's a bit irritating since it's converted to a more readable representation when the next key gets recorded or when the shortcut is complete)
* this is certainly part of the current concept - but should we also allow single key shortcuts or will that conflict too much with keyboard input when entering actual text?

umlaeute commented:
> personally, i admit that i barely ever change shortcuts for any application since i tend to hope for good defaults 

Same for me 😉
That's why I had trouble finding the relevant configurations in other programs (I knew of the GIMP (where the shortcut editor is rather prominent), but I only learned about inkscape and libreoffice last night)

> 
> no scrollbar visible for me

There is none  Is that a problem?
The panel is still scrollable for me, so I didn't see a real reason to add one


> touchpad movements stop the "recording" when the cursor is moved out off the recorded cell

This is somewhat intentional.
If it is too annoying, we can change it to some more explicit action to stop recording.


> the shortcut text is clipped a bit much in the cell while recording with fontsize 10

Hmm. 
You can resize the columns, but of course you shouldn't need to.

> modifiers appear as 2 keys while recording, like "Command+Meta_L" (i guess, this is a proper representation - but it's a bit irritating since it's converted to a more readable representation when the next key gets recorded or when the shortcut is complete)

Yes.
It's a bit complicated as the key's role switches from "terminal key" to "modifier", and they are repressed differently from the underlying TclTk.
The only way to fix this is would be to special cases each of those keys, and I would like to avoid this: special keys are already reported with different names under each windowing system, as are other nonstandard keys (have you tried "ö"?).
There is no telling how many different messages there are (I grepped xkeynames.h and it were beyond 1000)


> this is certainly part of the current concept - but should we also allow single key shortcuts or will that conflict too much with keyboard input when entering actual text?

That was my thinking. 
It's also the reason why <kbd>shift</kbd>+key (without another modifier) is not allowed. 
OTOH, we could of course allow the users to shoot themselves in their own knee at their own discretion (so you could map some never-used keys like <kbd>Print</kbd>).
Maybe the user has to tick a checkbox to allow single-key shortcuts as a safety measure.

porres commented:
I'd suggest you to check PlugData, which has shortcut edits

You can install it too ;) 

pretty intuitive and self-explanatory

<img width="789" alt="Screenshot 2024-07-15 at 04 04 54" src="https://github.com/user-attachments/assets/17bb7aa4-a86b-483c-8572-c37ca7d194ba">

<img width="536" alt="Screenshot 2024-07-15 at 04 05 09" src="https://github.com/user-attachments/assets/9b6fc383-5f87-4abc-afbc-3dd934def6d8">

umlaeute commented:
> I'd suggest you to check PlugData, which has shortcut edits

i would like to use an out-of-our-ecosystem tried-and-proofed model

umlaeute committed:
::pd_menucommands::menu_print: make sure that this acts on toplevel windows

umlaeute commented:
@porres write:
> The instructions need to say "To edit a shortcut, doubleclick on the corresponding ROW AND COLUMN", or "cell", or something. Just "row" is wrong and misleading.

i've changed the description to no longer speak of the underlying layout altogether, but instead use the actual meaning of those entries.
I've also decided to do away with using keystrokes to interact with the shortcut recorder, as this prevents using some keys for shortcuts (that is:  if <kbd>BackSpace</kbd> deletes a shortcut, you cannot use this key in a shortcut itself).
so this is somewhat similar to PlugData.

the new shorthelp is:
> To edit a keyboard shortcut, double-click on the corresponding event (or shortcut) and type a new accelerator.
> Use right-click for a context-menu to remove shortcuts.

i've also removed the "moving the mouse out of the recording-cell cancels" as suggested by @ben-wes , and you know have to *click* somewhere.

umlaeute committed:
pd_menus: map tcl keynames to pretty names

umlaeute committed:
pd_bindings: use events %W rather than $::focused_window

umlaeute committed:
integrate shortcut editor

into preferences and build-system

umlaeute committed:
use "Clear Recent Files" rather than "Clear Menu"

it makes more sense when your focus is not on the "Recent Files" menu

umlaeute commented:
rebased on master.

if somebody wants to still test ;-)

umlaeute committed:
::pd_menus::insert_into_menu compat with Tcl/Tk-8.5

[menu index end] returns "none" on some system  (like Tk8.5 on Windows),
presumably for empty menus), but '0' on other systems.

Ant1r commented:
This is great!
I've just used it to try adding bindings for `Put object/message/(...)/comment` to numeric keypad (BTW I wonder if this shouldn't be the default, after all...)

Ant1r commented:
sidenote: here on Linux the config is stored in `~/.config/KeyBindings.conf`, which is fine, but all the settings are grouped in one single line, totaling 2137 characters...
With a basic editor and an old computer, this is barely usable.

Would it be possible to put each setting in its own line? Even if we'll rarely have to actually edit this text by hand...

umlaeute commented:
> > not related to _this_ pr
> 
> This is the only conf file this long, so I don't think it's such a general issue.

that doesn't really matter.
*this* PR just passes a mapping list onto the [GUI-preferences](https://github.com/pure-data/pure-data/blob/master/tcl/pd_guiprefs.tcl). Where the preferences are stored (Windows registry, macOS plists, flat files on Linux) and how they are formatted within the storage is at the sole discretion of the GUI-preferences.

there is nothing this PR can (and should) do about it.

(and yes, afaik as of yet the keyboard shortcuts are the biggest data chunk that the GUI preferences need to store, that's why you see it only there)

umlaeute committed:
pd_bindings: use ${var} rather than $var

umlaeute committed:
pd_bindings: use $::pd_bindings::control rather than $::modifier

umlaeute committed:
pd_bindings: use "Command" for keybinding on macOS, rather than "Mod1"

it's really the same, but everyone knows what "Command" is

umlaeute committed:
sort "Message..." bindings under Tools
